src/main/java/com/faforever/client/fx/LineChartWithHover.java

```diff
@@ -0,0 +1,190 @@
+      });
+      addVerticalLine();
+    }
+  }
+
+  private void addVerticalLine() {
+    getPlotChildren().add(verticalLine);
+    JavaFxUtil.bind(verticalLine.endYProperty(), chartBackground.heightProperty());
+  }
+
+  public void setTooltipContainer(LineChartTooltip tooltipContainer) {
+    if (available) {
+      this.tooltipContainer = tooltipContainer;
+      prepareHoverTooltip();
+    }
+  }
+
+  private void prepareHoverTooltip() {
+    hoverTooltip = JavaFxUtil.createCustomTooltip(tooltipContainer.getRoot());
+    hoverTooltip.setAutoFix(false);
+    hoverTooltip.setHideDelay(Duration.ZERO);
+    hoverTooltip.setShowDelay(Duration.ZERO);
+    hoverTooltip.setShowDuration(Duration.INDEFINITE);
+    Tooltip.install(chartBackground, hoverTooltip);
+  }
+
+  private void showCurrentValues(MouseEvent event) {
+    if (valid) {
+      double x = event.getX();
+      Double xValue = getDisplayedXValue(x);
+      Double yValue = yValuesMap.get((int) x);
+      if (yValue != null) {
+        tooltipContainer.currentXY(xValue, yValue);
+      } else {
+        tooltipContainer.empty();
+      }
+      moveTooltipAndLineToCursor(event);
+    }
+  }
+
+  private void moveTooltipAndLineToCursor(MouseEvent event) {
+    hoverTooltip.setX(event.getScreenX() + OFFSET_TOOLTIP_FROM_CURSOR);
+    hoverTooltip.setY(event.getScreenY() + OFFSET_TOOLTIP_FROM_CURSOR);
+    verticalLine.setLayoutX(event.getX());
+  }
+
+  private void recalculateData() {
+    yValuesMap.clear();
+    for (int i = 0; i < getData().size(); i++) {
+      Series<X, Y> series = getData().get(i);
+      if (series.getNode() instanceof Path) {
+        buildData(((Path) series.getNode()).getElements());
+        break;
+      }
+    }
+  }
+
+  private void buildData(List<PathElement> elements) {
+    List<LineTo> lines = elements.stream().filter((e) -> e instanceof LineTo).map((e) -> (LineTo) e).collect(Collectors.toList());
+    if (lines.size() < 2) {
+      return;
+    }
+
+    for (int i = 0; i < lines.size() - 1; i++) {
+      LineTo line1 = lines.get(i);
+      LineTo line2 = lines.get(i + 1);
+      int x1 = (int) Math.floor(line1.getX());
+      int x2 = (int) Math.floor(line2.getX());
+      Number yValue1 = getDisplayedYValue(line1.getY());
+      Number yValue2 = getDisplayedYValue(line2.getY());
+      int distance = x2 - x1;
+      if (distance > 1) {
+        double k = (yValue2.doubleValue() - yValue1.doubleValue()) / distance;
+        double yValue = yValue1.doubleValue();
+        while (x1 < x2) {
+          putYValueMap(++x1, yValue += k);
+        }
+      } else {
+        putYValueMap(x1, yValue1);
+      }
+    }
+  }
+
+  private void putYValueMap(int x, Number y) {
+    Double value = yValuesMap.get(x);
+    yValuesMap.put(x, value != null
+        ? (value + y.doubleValue()) / 2
+        : y.doubleValue());
+  }
+
+  private Double getDisplayedXValue(double displayPosition) {
+    return (Double) getXAxis().getValueForDisplay(displayPosition);
```

Comment:
This casting will only work when the axis is of double objects so this class may only be valid for when X and Y are doubles